### PR TITLE
[tune/ci] Multinode support killing nodes in Ray client mode

### LINKED ISSUE #21654: [fake multi node] provide a mechanism to kill node under ray client mode
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/ray/issues) and found no similar feature requirement.

### Description

`docker kill <node>` can only work from outside of the docker nodes. However, tune Callbacks are called from driver (which runs on head node). We need a mechanism for tune Callbacks to surface the information to outside of the docker compose cluster, where `docker kill` can be invoked, presumably through `ray.queue`.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

The multi node testing utility currently does not support controlling cluster state from within Ray tasks or actors., but it currently requires Ray client. This makes it impossible to properly test e.g. fault tolerance, as the driver has to be executed on the client machine in order to control cluster state. However, this client machine is not part of the Ray cluster and can't schedule tasks on the local node - which is required by some utilities, e.g. checkpoint to driver syncing.

This PR introduces a remote control API for the multi node cluster utility that utilizes a Ray queue to communicate with an execution thread. That way we can instruct cluster commands from within the Ray cluster.

## Related issue number

Closes #21654

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by xwjiang2010 — APPROVED** [review]
Thanks Kai! Some nits.

Another thing is I feel like we should make fake multi node setup to support non ray client mode as its MVP use case. ray client mode should be a good to have.

File: python/ray/autoscaler/_private/fake_multi_node/test_utils.py [hunk 1]
```diff
@@ -340,3 +367,28 @@ def kill_node(
         node_id = self._get_node(node_id=node_id, num=num, rand=rand)
         container = self._get_docker_container(node_id=node_id)
         subprocess.check_output(f"docker kill {container}", shell=True)
+
+
+class RemoteAPI:
```
[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/autoscaler/_private/fake_multi_node/test_utils.py:372]: It's returned by `DockerCluster.remote_execution_api()` - I've added a comment that it's serializable, can be passed to Ray actors but also be used locally.

File: python/ray/autoscaler/_private/fake_multi_node/test_utils.py [hunk 2]
```diff
@@ -116,6 +122,26 @@ def connect(self, client: bool = True, timeout: int = 120, **init_kwargs):
...
+
+        def entrypoint():
+            while not stop_event.is_set():
+                try:
+                    cmd, kwargs = self._execution_queue.get(timeout=1)
+                except Empty:
+                    continue
+
+                if cmd == "kill_node":
+                    self.kill_node(**kwargs)
```
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/autoscaler/_private/fake_multi_node/test_utils.py:138]: well, I guess GIL guarantees that?

[the remark this answers is not in the document]
Comment (krfricke) [reply python/ray/autoscaler/_private/fake_multi_node/test_utils.py:138]: It won't be an issue for calling the methods, but we can definitely run into race conditions here. `get_node()` will read node state from a json file, which is updated by an external monitoring script. Killing a node will impact this as soon as the docker daemon actually killed the node.

In short, we have to be careful here :-)

**Comment by krfricke** [comment]
Yes I agree, we should aim for non-client mode next. This utility was just blocking the syncer refactor so I got to it first.

**Pull request merged**